servicetalk-transport-netty-internal/src/main/java/io/servicetalk/transport/netty/internal/RequestResponseCloseHandler.java

```diff
@@ -162,7 +179,7 @@ public void protocolPayloadBeginOutbound(final ChannelHandlerContext ctx) {
 
     @Override
     public void protocolPayloadEndOutbound(final ChannelHandlerContext ctx) {
-        if (isClient || (has(state, CLOSING) && pending == 0)) {
+        if (isClient || (closeEvent != null && pending == 0)) {
             ctx.pipeline().fireUserEventTriggered(OutboundDataEndEvent.INSTANCE);
```

Comment:
not introduced by this PR but the control flow here is somewhat confusing. This method is called before the trailers are actually written, and eventually goes to `WriteStreamSubscriber#channelOutboundClosed` which calls `sourceTerminated`, which will complete the promise if `activeWrites==0` but since we haven't written yet I assume that is never possible?

Reply:
Agreed, the flow here is confusing. It was originally introduced in #919 and later improved (there was a bug) in #1155, according to the original logic from #919.
IIUC this was done specifically for h2. The event is emitted before write, because netty closes the stream in write listener. But not sure if it need to be done in the same way for h1. I had an idea of how to simplify it by merging `protocolPayloadEndOutbound` and `protocolPayloadEndOutboundSuccess` together. Implementations of CloseHandler for h1 and h2 are anyway different. But did not have time to experiment. Decided to stop introducing more concurrent PRs for the same part of the codebase.